### PR TITLE
feat(gradients): add ivy.vjp and ivy.jvp

### PR DESCRIPTION
**Pull request by yopknopixx**
# PR Description

This PR adds support for `ivy.vjp` and `ivy.jvp` per [this Trello task](https://trello.com/c/JJDL7MYZ/1485-implement-ivyjvp-and-ivyjvp).

## Checklist

- [x] Implement `ivy.jvp` for [`jax`, `torch`, `tensorflow`, `mxnet`, `paddle`]
- [x] Implement `ivy.vjp` for [`jax`, `torch`, `tensorflow`]
- [x] Add tests
- [ ] Refactor to support containers

### CODE REVIEW

**Comment by ivy-leaves** [comment]
If you are working on an open task, please **edit** the PR description to link to the issue you've created.

For more information, please check [ToDo List Issues Guide](https://unify.ai/docs/ivy/overview/contributing/the_basics.html#todo-list-issues).

Thank you :hugs:

File: ivy/functional/backends/torch/experimental/gradients.py [hunk 1]
```diff
@@ -25,3 +26,36 @@ def backward(ctx, upstream):
 
     custom_module = _CustomModule.apply
     return inputs_to_native_arrays(custom_module)
+
+
+def vjp(func: Callable, *primals):
```
Comment (vedpatwardhan) [inline ivy/functional/backends/torch/experimental/gradients.py:31]: For `torch` do you think we could use something like [`torch.autograd.functional.vjp`](https://pytorch.org/docs/stable/generated/torch.autograd.functional.vjp.html)?

Comment (vedpatwardhan) [reply ivy/functional/backends/torch/experimental/gradients.py:31]: same with `torch.autograd.functional.jvp`

**Review by vedpatwardhan — CHANGES_REQUESTED** [review]
Hey @yopknopixx, just requested a minor change, rest seems good to merge. Thanks 😄

File: ivy/functional/backends/torch/experimental/gradients.py (hunk 1 above)
Comment (yopknopixx) [reply ivy/functional/backends/torch/experimental/gradients.py:31]: That would've been ideal, but `torch.autograd.functional.vjp` requires the `v` (`tangents` in `jax`) argument to be passed for `size` > 1. That doesn't align with the `jax` API where `tangents` is passed to a callback function returned by `vjp`.
We could do an explicit computation for `func_output` then call `vjp` in the callback function if we wanted to call from `torch.autograd.functional`. But that will have us computing `func_output` twice.
What do you think?

Comment (vedpatwardhan) [reply ivy/functional/backends/torch/experimental/gradients.py:31]: actually `jax.vjp` returns a callback whereas `torch.autograd.functional.vjp` returns the product computed. So given that we need to pass `v` to the callback anyway for computing the product, an ideal comparison would be between `jax.vjp(func, x)(v)` with `torch.autograd.functional.vjp(func, x, v)` instead of `jax.vjp(func, x)`. 
Based on this I think both `jax.vjp` and `torch.autograd.functional.vjp` require `v` anyway, so we could just wrap this logic into a function and return that function from `ivy.vjp`
e.g. for the `torch` backend, it could be something like,
```
def vjp(func, *inputs):
    def vjp_func(v):
        return torch.autograd.functional.vjp(func, inputs, v)
    return vjp_func
```
and then for `jax`, we'd just return `jax.vjp(func, *inputs)` if that makes sense?
Happy to clarify further if required, I could be missing something.

Comment (yopknopixx) [reply ivy/functional/backends/torch/experimental/gradients.py:31]: That makes sense, but both `torch` and `jax` return the function outputs when calling `vjp`. So what it would look like in the `torch` backend is:
```
def vjp(func, *inputs):
    outputs = func(*inputs)
    def vjp_func(v):
        outputs, vjp = torch.autograd.functional.vjp(func, inputs, v)
        return vjp
    return (outputs, vjp_func)

```
As you can see, we'd be calculating the outputs twice. If that is fine, we can go ahead with `torch.autograd.functional.vjp`.

Comment (vedpatwardhan) [reply ivy/functional/backends/torch/experimental/gradients.py:31]: Ah yeah you're right, sorry missed that part! What do you think about adding `v` as an argument to `ivy.vjp` then? Probably that's the only way we can get all backends work consistently without doing the computation twice

Comment (yopknopixx) [reply ivy/functional/backends/torch/experimental/gradients.py:31]: That would work for consistency (with the current `paddle` implementation we are computing the outputs twice) but the problem I see with that is this:
When a user wants to do backward mode autodiff, they expect to do something like this:
```
x1, back_x1 = vjp(layer1, x)
x2, back_x2 = vjp(layer2, x1)
# calculate gradients (assuming x2 is scalar)
grad = back_x1(back_x2(tensor(1.)))
```
So, they won't have the `v` argument ready till the next computation is done. This would force them to compute the outputs twice, once during the forward propagation and again when calling `vjp` to compute gradients.

Comment (vedpatwardhan) [reply ivy/functional/backends/torch/experimental/gradients.py:31]: Yeah but if `v` is an argument accepted in the `vjp` function anyway, wouldn't we simply be returning the product instead of a callback? I mean if we have everything we need to compute the product inside the `ivy.vjp` function itself, it wouldn't really be required to return a callback that accepts the `v` (just like `torch.autograd.functional.vjp`). Let me know what you think 😄

Comment (yopknopixx) [reply ivy/functional/backends/torch/experimental/gradients.py:31]: I understand your point, and it makes total sense from `ivy` PoV. But, from user PoV, consider the original use case of `thinc` that needed this function:
```
    def begin_update(self, inputs: ArgsKwargs):
    output, vjp_fn = ivy.vjp(
        lambda v, x: self._model(x, v=v),
        self._model.v,
        inputs.args[0],
    )

    def backprop(external_grads):
        self.param_grads = vjp_fn(external_grads)
        return self.param_grads

    return output, backprop
```
In this case, and many other cases like this, the user won't have the `v` argument ready. And if they have to use `ivy.vjp(func, inputs, v)` it would go something like this:
```
def begin_update(self, inputs: ArgsKwargs):
    output = self._model(*inputs.args, **inputs.kwargs) # 1st computation of outputs

    def backprop(external_grads):
        outputs, self.param_grads = ivy.vjp(self._model, inputs, external_grads) # 2nd computation of outputs
        return self.param_grads

    return output, backprop
```
So, the user would have to calculate the outputs twice. That's why I think having a callback is necessary. Let me know what you think about that.
Regardless, if using `torch.autograd.functional.vjp` is important from the compilation perspective, We can include the `v` argument in `vjp` directly 😊.

Comment (vedpatwardhan) [reply ivy/functional/backends/torch/experimental/gradients.py:31]: Yeah you're right, that makes sense! Given that our gradients API is functional we can't really track operations as they're executed with `torch` as a result of which `torch.autograd.functional.vjp` in `ivy.vjp` by adding the `v` parameter doesn't make sense unless we do some weird wrapping of the product itself (`lambda: product`) or something when calling `ivy.vjp` and then use that callback inside the function defined in your example to avoid computing twice. 
Yeah let's not use `torch.autograd.function.vjp` and mention the reason behind doing that as a comment in the `torch` backend implementation for future reference. 
Thanks for the deep dive @yopknopixx, I'll resolve this conversation now 😄

**Review by vedpatwardhan — APPROVED** [review]
lgtm! Feel free to merge, thanks @yopknopixx 😄
